src/lib.rs

```diff
@@ -0,0 +1,489 @@
+// if one is included with the Software (each a "Larger Work" to which the Software
+// is contributed by such licensors),
+//
+// without restriction, including without limitation the rights to copy, create
+// derivative works of, display, perform, and distribute the Software and make,
+// use, sell, offer for sale, import, export, have made, and have sold the
+// Software and the Larger Work(s), and to sublicense the foregoing rights on
+// either these or other terms.
+//
+// This license is subject to the following condition: The above copyright
+// notice and either this complete permission notice or at a minimum a reference
+// to the UPL must be included in all copies or substantial portions of the
+// Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#[macro_use] extern crate lazy_static;
+#[macro_use] extern crate log;
+extern crate env_logger;
+
+use std::collections::HashMap;
+use std::sync::RwLock;
+use std::thread;
+use std::thread::JoinHandle;
+use std::cell::RefCell;
+
+
+/// Uniquely identifies a location in the user-program.
+///
+/// XXX Hard-coded to a simple program counter value.
+/// XXX This need to be user-specified, perhaps using a trait?
+type JitLocation = u32;
+
+/// The type used to count how many times a JitLocation has been interpreted.
+type HotCount = u32;
+
+/// The number of times a location is interpreted before we trace it.
+/// XXX make this configurable.
+const HOT_THRESHOLD: HotCount = 10;
+
+/// The shared trace store.
+///
+/// Maps a location to a trace (if one exists). Locations not in the map have not yet been traced.
+lazy_static! {
+    static ref TRACES: RwLock<HashMap<JitLocation, JitTrace>> = RwLock::new(HashMap::new());
+}
+
+/// Execute the trace for the specified location (if there is one).
+///
+/// Returns true if a trace was present (and thus was executed), or false otherwise.
+fn try_exec_trace(loc: JitLocation) -> bool {
+    let trace_map = TRACES.read().unwrap();
+    let trace = trace_map.get(&loc);
+    if trace == Some(&JitTrace::FinalisedTrace) {
+        JitThreadContext::phase_transition(JitPhase::RunningTrace(loc));
+        // XXX Really execute the trace.
+        JitThreadContext::phase_transition(JitPhase::Interpreting);
+        return true;
+    }
+    false
+}
+
+/// Is a thread currently tracing/compiling the specified location?
+fn trace_pending(loc: JitLocation) -> bool {
+    let trace_map = TRACES.read().unwrap();
+    match trace_map.get(&loc) {
+        Some(&JitTrace::TracePending) => return true,
```

Comment:
If you remove this `return` and the one on the line below, you'll end up with semantically equivalent (and more Rusty) code.

Reply:
That statement has never been true, because Rust has always been an expression-based language ;)

Reply:
I could have swore:
```
let ret = match {...};
ret
```

was an idiom due to some restriction at some point.

None the less, it works now:
https://play.rust-lang.org/?gist=bbd15f3fed27a84c71e7b4036519e218&version=stable

Reply:
Rust has always been an expression based language, so this has always worked. You've been swearing for no good reason ;)